Add CI Tests for test infra and partner image

mkochanowski opened the pull request:
Implements three new GitHub Workflows for testing, and a GitHub Action for setting up a minikube cluster suitable for CI/CD pipelines.

The new `start-minikube` GitHub Action should be considered as the recommended Action for creating test clusters for TNF resources in CI workflows, removing the dependency on third-party, untrusted Actions from the GitHub Marketplace.

The new workflows are:

- `cnf-test-partner-image.yaml`: Tests whether the partner image can be built, and whether the image contains all required tools.

- `local-test-infra.yaml`: Tests the OpenShift resources defined in the local-test-infra directory.
  The resources are created on the minikube test cluster using commands exposed in the Makefile.

- `partner-pod.yaml`: Tests the installation process of the partner pod according to the instructions in the README. 
  The pod is created on the minikube test cluster.

Signed-off-by: Marek Kochanowski <mkochanowski@redhat.com>

crwr45 reviewed (approved):
Excellent work

ryandgoulding commented:
Why were no "checks" run as a part of this PR?  The checks should have automatically run, which makes me believe the checkout action isn't working properly here.

ryandgoulding commented on .github/actions/start-minikube/action.yaml:
```diff
@@ -0,0 +1,47 @@
+runs:
+  using: 'composite'
+  steps:
+    - name: Download minikube debian package
+      run: |
+        curl -Lo minikube.deb https://storage.googleapis.com/minikube/releases/${{ inputs.minikube_version }}/minikube_latest_amd64.deb
+      shell: bash
+
+    - name: Install minikube
+      run: sudo dpkg -i minikube.deb
+      shell: bash
+
+    - name: Set the default VM driver
+      run: minikube config set vm-driver ${{ inputs.minikube_driver }}
+      shell: bash
+
+    - name: Start the minikube cluster
+      run: minikube start ${{ inputs.minikube_extra_options }}
+      shell: bash
+
+    # The waiting is performed to allow the `kube-controller-manager` to start all required controllers,
+    # such as the `service-account-controller` responsible for managing and creating Service Accounts.
+    - name: Wait for kube-controller-manager
+      run: |
+        echo "Waiting for kube-controller-manager... (timeout: ${{ inputs.oc_kcm_timeout }})" && \
+        oc wait -n kube-system --for=condition=ready pod/kube-controller-manager-minikube --timeout=${{ inputs.oc_kcm_timeout }}
+      shell: bash
+
+    - name: Display kube-system pods
+      run: oc get pods -n kube-system -o wide
+      shell: bash
+
+    - name: Display a friendly summary message
+      run: echo "TNF-compatible minikube cluster is ready to go! Have fun!"
+      shell: bash
```
May be helpful to pull this out for reuse somewhere else.  Not necessary for this patch though!

ryandgoulding replied:
Actually, nevermind.  It should probably stay here for right now.  Standing up an additional repository is not worth it for a single GitHub Action.  Great job!

mkochanowski replied:
Isolating the Action so it can be reused later was the intention!

For now, if we want to use this in the main TNF repository's workflows, we can call the `actions/checkout@v2` to get this action included where it's needed.

mkochanowski commented:
> Why were no "checks" run as a part of this PR? The checks should have automatically run, which makes me believe the checkout action isn't working properly here.

Each workflow definition is equipped with path filters, like this: 
```yaml
name: Test the `cnf-test-partner` image
on:
  push:
    paths:
      - 'test-partner/Dockerfile'
```
The filter ensures that the workflow is run if any changes are to be introduced to the selected parts of the codebase (in this example, if the Dockerfile is modified).

For the `cnf-test-partner` image, I think we would like to test and auto-push (TBD) a new container image only in the situation when changes are introduced to the Dockerfile, rather than doing it for every PR (e.g. fixing a typo in the documentation should not trigger the auto-push).

In most CI systems the filtering is done to reduce the number of concurrent jobs. 
In extreme cases, if the number of concurrent jobs permitted at a time is very low or the volume of contributions is high (let's say, everyone has the same working hours), running all workflows on every commit can get them queued up, clogging the CI pipeline for many minutes or hours, for every contributor in the organization. 
It's probably a premature optimization, but a good practice nonetheless. I couldn't find what's the exact number of allowed concurrent jobs on GitHub Actions for the free tier (20 per org?).

Possible actions include:
1. Pushing a test commit to this branch that contains sample changes. This will trigger all workflows. The test commit can then be removed before merging.
2. Removing the path filters in all workflows, excluding the `cnf-test-partner` image workflow. This will result in the the container image being built and tested only for PRs that modify the Dockerfile.
3. Removing the path filters in all workflows. This will result in all checks being triggered for every commit and PR, no matter the contents of the PR.

I think the 2. option might be good enough for our needs. Any thoughts?

*Note*: the `test-partner-pod.yaml` check will fail until we merge the PR that adds the `test-partner/namespace.yaml` resource.

ryandgoulding commented:
@mkochanowski option 2 seems correct.

mkochanowski commented:
@ryandgoulding I'll update the branch with the option 2 then.

However, it seems like no matter which option we would have gone with, GitHub prevents the execution of workflows supplied in a Pull Request from a forked repository, and there does not seem to be a setting that alters this behavior (maybe for the better).

**Reason**: Assume a bad actor that submits a malicious Pull Request containing new or updated workflows. 
Let the new workflow mine bitcoin. 
If the organization is on a paid plan, or uses self-hosted runners, the workflow would be executed against the organization's infrastructure, and would consume the organization's quota.
This may result in the organization receiving an unexpected, large bill to settle.

**Explanation**: When a new Pull Request is opened, a `pull_request` event is triggered on the _base repository_, and not on the _fork_. The `pull_request` event applies to workflows that are _already present_ in the repository. 
For this Pull Request, there are no workflows present in the base repository.

When possible, GitHub prefers to execute workflows on the _fork_. This mostly applies to `push` events; does not apply to `pull_request` (unless a PR is opened against the _fork_ itself).
The workflows ran on the _fork_ are not displayed on the PR page, only the workflows executed on the _base repository_ are.

Source: https://github.community/t/run-a-github-action-on-pull-request-for-pr-opened-from-a-forked-repo/16054

mkochanowski commented:
The branch is now updated with path filters being removed in all workflows for all event types, except for the `push` event of `cnf-test-partner-image.yaml`. 
This workflow is responsible for building, testing and pushing (#12) the image. The workflow only applies to commits affecting the Dockerfile.

I've also removed the `test-*` prefix from filenames, as we won't be creating separate workflows (like `test-*`, `push-*`) for the same resource type (the `workflow_run` event type is currently too limiting to make this viable).
